example/contrib/socks-client/socks/impl/handshake.hpp

```diff
@@ -0,0 +1,598 @@
+//
+// Copyright (c) 2018 jackarain (jack dot wgm at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/boostorg/beast
+//
+
+#ifndef SOCKS_IMPL_HANDSHAKE_HPP
+#define SOCKS_IMPL_HANDSHAKE_HPP
+
+#include <socks/detail/protocol.hpp>
+#include <boost/beast/core/async_base.hpp>
+#include <boost/beast/core/detail/is_invocable.hpp>
+#include <boost/config.hpp> // for BOOST_FALLTHROUGH
+#include <boost/asio/ip/tcp.hpp>
+#include <boost/asio/streambuf.hpp>
+#include <boost/asio/read.hpp>
+#include <boost/asio/write.hpp>
+#include <boost/asio/buffer.hpp>
+#include <iostream>
+#include <string>
+#include <memory>
+#include <utility>
+#include <type_traits>
+
+
+namespace socks {
+
+namespace detail {
+
+template<typename type, typename source>
+type read(source& p)
+{
+    type ret = 0;
+    for (std::size_t i = 0; i < sizeof(type); i++)
+        ret = (ret << 8) | (static_cast<unsigned char>(*p++));
+    return ret;
+}
+
+template<typename type, typename target>
+void write(type v, target& p)
+{
+    for (auto i = (int)sizeof(type) - 1; i >= 0; i--, p++)
+        *p = static_cast<unsigned char>((v >> (i * 8)) & 0xff);
+}
+
+} // detail
+
+template<
+    class Stream,
+    class Handler,
+    class Buffer,
+    class base_type = boost::beast::async_base<
+        Handler, typename Stream::executor_type>>
+class socks4_op : public base_type
+{
+public:
+    socks4_op(socks4_op&&) = default;
+    socks4_op(socks4_op const&) = default;
+
+    socks4_op(Stream& stream, Handler& handler,
+        const std::string& hostname, unsigned short port,
+        const std::string& username)
+        : base_type(std::move(handler), stream.get_executor())
+        , stream_(stream)
+        , hostname_(hostname)
+        , port_(port)
+        , username_(username)
+    {
+        (*this)({}, 0); // start the operation
+    }
+
+    void
+    operator()(
+        error_code ec,
+        std::size_t bytes_transferred)
+    {
+        boost::ignore_unused(bytes_transferred);
+
+        using detail::write;
+        using detail::read;
+
+        auto& response = *response_;
+        auto& request = *request_;
+
+        switch (ec ? 3 : step_)
```

Comment:
Instead of `switch` and `step_`, this should use `<boost/asio/coroutine.hpp>` and the macros `BOOST_ASIO_CORO_REENTER`, `BOOST_ASIO_CORO_YIELD`, it will make the code easier to reason about.